File: medium-and-easy/947-most-stones-removed-with-same-row-or-column/Solution947.js
/**
 * @param {number[][]} stones
 * @return {number}
 */

class UnionFind {
    constructor() {
        this.parents = [];
        this.movesCount = 0;
    }

    find(x) {

        let rootParent = x;
        while (rootParent != this.parents[rootParent]) {
            rootParent = this.parents[rootParent];
        }

        let temp,
            current = x;

        while (current != this.parents[current]) {
            temp = this.parents[current];
            this.parents[current] = rootParent;
            current = temp;
        }
        return rootParent;
    }

    union(a, b) {
        let rootA = this.find(a);
        let rootB = this.find(b);
        this.parents[rootB] = rootA;
    }

    test() {
        console.log(`just a test function`);
    }

}


var removeStones = function(stones) {

    let uf = new UnionFind();

    let maxRow = -1,
        maxCol = -1;

    stones.forEach(function(stone) {
        maxRow = Math.max(stone[0] + 1, maxRow);
        maxCol = Math.max(stone[1] + 1, maxCol);
    });

    for (var i = 0; i < maxRow + maxCol; i++) {
        uf.parents[i] = i;
    }

    stones.forEach((stone) => {
        uf.union(stone[0], stone[1] + maxRow);
    });

    let parentsSeen = new Set();

    stones.forEach((stone) => {
        parentsSeen.add(uf.find(stone[0]));
    })

    return stones.length - parentsSeen.size;

};

File: medium-and-easy/947-most-stones-removed-with-same-row-or-column/Solution947.java
import java.util.HashSet;
import java.util.Set;

class Solution947 {
    
    public class UnionFind {
        
        int[] parent;
        
        public UnionFind(int n) {
            parent = new int[n];
            for (int i = 0; i < n; i ++) {
                parent[i] = i;
            }
        }
        
        public int find(int i) {
            if(parent[i] != i) {
                parent[i] = find(parent[i]);
            }
            return parent[i];
        }
        
        public void union(int a, int b) {
            
            int parentA = find(a);
            int parentB = find(b);
            
            parent[parentB] = parentA; 
        }
        
    }
    
    public int removeStones(int[][] stones) {
        
        if (stones.length == 0 || stones.length == 1) return 0;
        
        int n = stones.length;
        UnionFind uf = new UnionFind(n);
        
        
        for (int i = 0; i < n; i ++) {
            for (int j = i + 1; j < n; j++) {
                if(stones[i][0] == stones[j][0] || stones[i][1] == stones[j][1]) {
                    uf.union(i, j);
                }
                
            }
        }
        
        Set<Integer> seen = new HashSet<Integer>();
        for (int i = 0; i < n; i++) {
            seen.add(uf.find(i));
        }
        // System.out.println(seen.size());
        
        return n - seen.size();
        
    }
}